lib/improver/utilities/solar.py

```diff
@@ -0,0 +1,259 @@
+
+        Returns:
+            daynight_mask (iris.cube.Cube):
+                Blank daynight mask cube. The resulting cube will be the
+                same shape as the time, y, and x coordinate, other coordinates
+                will be ignored although they might appear as attributes
+                on the cube as it is extracted from the first slice.
+        """
+        daynight_mask = next(cube.slices([cube.coord('time'),
+                                          cube.coord(axis='y'),
+                                          cube.coord(axis='x')])).copy()
+        daynight_mask.long_name = 'day_night_mask'
+        daynight_mask.standard_name = None
+        daynight_mask.var_name = None
+        daynight_mask.units = 1
+        daynight_mask.data = np.ones(daynight_mask.data.shape,
+                                     dtype='int')*self.NIGHT
+        return daynight_mask
+
+    def _daynight_lat_lon_cube(self, mask_cube, day_of_year, utc_hour):
+        """
+        Calculate the daynight mask for the provided Lat Lon cube
+
+        Args:
+            mask_cube (iris.cube.Cube):
+                daynight mask cube - data initially set to self.NIGHT
+            day_of_year (int):
+                day of the year
+            utc_hour (float):
+                Hour in UTC
+
+        Returns:
+            mask_cube (iris.cube.Cube):
+                daynight mask cube - daytime set to self.DAY
+        """
+        lons = mask_cube.coord('longitude').points
+        lats = mask_cube.coord('latitude').points
+        terminator_lats = daynight_terminator(lons, day_of_year, utc_hour)
+        lons_zeros = np.zeros_like(lons)
+        lats_zeros = np.zeros_like(lats).reshape(len(lats), 1)
+        lats_on_lon = lats.reshape(len(lats), 1) + lons_zeros
+        terminator_on_lon = lats_zeros + terminator_lats
+        dec = solar_declination(day_of_year)
+        if dec > 0.0:
+            index = np.where(lats_on_lon >= terminator_on_lon)
+        else:
+            index = np.where(lats_on_lon < terminator_on_lon)
+        mask_cube.data[index] = self.DAY
+        return mask_cube
+
+    def process(self, cube):
+        """
+        Calculate the daynight mask for the provided cube
+
+        Args:
+            cube (iris.cube.Cube):
+                input cube
+
+        Returns:
+            daynight_mask (iris.cube.Cube):
+                daynight mask cube, daytime set to self.DAY
+                nighttime set to self.NIGHT.
+                The resulting cube will be the same shape as
+                the time, y, and x coordinate, other coordinates
+                will be ignored although they might appear as attributes
+                on the cube as it is extracted from the first slice.
+        """
+        daynight_mask = self._create_daynight_mask(cube)
+        dtvalues = iris_time_to_datetime(daynight_mask.coord('time'))
+        for i in range(0, len(daynight_mask.coord('time').points)):
```

Comment:
This is a bit nicer and allows you to get rid of the next two statements:
`for mask_cube, dtval in zip(daynight_mask, dtvalues):`

Reply:
I like this but I use i further down